.github/workflows/spread-test.yaml

```diff
@@ -4,19 +4,124 @@ on: [pull_request]
 
 jobs:
   spread:
-    runs-on: ubuntu-20.04
+    runs-on: self-hosted
     strategy:
       matrix:
         system:
           - ubuntu-18.04
           - ubuntu-20.04
+          - ubuntu-22.04
```

Comment:
Perhaps you could avoid using a matrix for this project. The matrix is usefull when you have a very long list of tests and the logs become very difficult to read. 
In this case if you use matrix, each job will be executed in a different runner and this project has 3 runners available. So, if you use the matrix, you just are able to run 1 pr in gitbug runners, if you dont use the matrix, you can run up to 3 pr in parallel.

Reply:
Also if you don't use matrix this file will be more simple as you can get rid of all the ${{ matrix.system }} in the code.
i.e.:

key: "${{ github.job }}-results-${{ github.run_id }}-${{ matrix.system }}-${{ steps.get-previous-attempt.outputs.previous_attempt }}" -> key: "${{ github.job }}-results-${{ github.run_id }}-${{ steps.get-previous-attempt.outputs.previous_attempt }}"

RUN_TESTS="$BACKEND:${{ matrix.system }}:tests/..." -> RUN_TESTS="$BACKEND:tests/..."